decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+
+#include <memory>
+#include <string>
+#include "../error_callback_fn.h"
+#include <thread>
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+			api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt);
+			return error_code::background_queue_overflow;
+		}
+
+	private:
+		void timer()//the timer triggers a queue flush (run in background)
+		{
+			while (_thread_is_running)
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(_batch_timeout_ms));
+				flush();
+			}
+		}
+
+		void flush()//flush all batches
+		{
+		  const auto queue_size = _queue.size();
+			if (queue_size == 0) return;
+
+			//handle batching
+			std::string batch, next;
+			_queue.pop(&batch); //there is at least one element
+
+			for (size_t i = 1; i < queue_size; ++i)
```

Comment:
>1 [](start = 19, length = 1)

why start at 1?

Reply:
Because the  `for` loop must iterate `queue_size-1` times, in order to dequeue all remaining elements
(The first one is dequeued before the loop, doing `_queue.pop(&batch)`.)

Reply:
ahhh... thanks!

---
In reply to: [186785374](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186785374) [](ancestors = 186785374)